Feature/6 0/asq api change proposal

mstuder opened the pull request:
Dear TB

This is our proposal for the new styled public api Assessment Question Service.

Please have a look to the Readme [Services/AssessmentQuestion/docs/readme.md](new UserAnswerSubmit(https://github.com/ILIAS-eLearning/ILIAS/blob/5784e0a1b78a56cb5751d93dc824e6ffa3b51a24/Services/AssessmentQuestion/docs/readme.md)

We are looking forward to the future discussion!!

Best regards
Björn & Martin

mstuder committed:
WIP Readme

mstuder committed:
WIP

mstuder committed:
WIP

AluethII committed:
first draw of asqauthoringservice

AluethII committed:
add skeleton for asq play service

AluethII committed:
add editor component skeleton

bheyser committed:
examples modified for new concept: ctrl structure forward, list questions using a table prvided by th econsumer

bheyser committed:
Merge branch 'feature/6-0/asq_api_change_proposal' into feature/6-0/asq_api_change_proposal_bh

bheyser committed:
examples modified for new concept: delete question from pool, renamed authoring spec things

mstuder committed:
Merge branch 'feature/6-0/asq_api_change_proposal_bh' into 'feature/6-0/asq_api_change_proposal'

Changes to the Examples from the original (historical) Asq Pull Request

See merge request ILIAS/Core!235

mstuder committed:
WIP

mstuder committed:
WIP

mstuder committed:
WIP

mstuder committed:
WIP

mstuder committed:
WIP

mstuder committed:
WIP

bheyser committed:
Merge branch 'feature/6-0/asq_api_change_proposal' into feature/6-0/asq_api_change_proposal_bh

mstuder committed:
WIP

mstuder committed:
WIP

bheyser committed:
Merge branch 'feature/6-0/asq_api_change_proposal' into feature/6-0/asq_api_change_proposal_bh

bheyser committed:
DIC / Factories

bheyser committed:
changes to factory due to renamed classes/interfaces/contracts and all the things

mstuder committed:
Merge branch 'feature/6-0/asq_api_change_proposal_bh' into 'feature/6-0/asq_api_change_proposal'

Feature/6 0/asq api change proposal bh

See merge request ILIAS/Core!238

bheyser committed:
export/import adapter connecting to the regular export/import service of ILIAS

bheyser committed:
fixed things

bheyser committed:
fixed things

bheyser committed:
changed example implementations for exObjQuestionPoolGUI

Commit:
removed AsqApi prefix

Commit:
few smaller changes, some phpdocs

mstuder committed:
Merge branch 'feature/6-0/asq_api_change_proposal_bh' into 'feature/6-0/asq_api_change_proposal'

Feature/6 0/asq api change proposal bh

See merge request ILIAS/Core!239

bheyser committed:
changed type hints in factories: Contacts Interfaces instead of ImplementationClasses

bheyser committed:
fixed ctrl gui class instantiation and fixed type hintings to contractInterface

bheyser committed:
removed useless namespace use

bheyser committed:
Question/RevisionId implementation, moved nix from revision to unique question id valobj, factory for identifiers

bheyser committed:
renamed some stuff another time

bheyser committed:
nearly completed pool example

bheyser committed:
completed pool table example

bheyser committed:
completed import/export examples

bheyser committed:
learning module export example finished

bheyser committed:
learning module export example finished

bheyser committed:
finally finished pool example with additional tax section for qst config forms

bheyser committed:
changed back to uuid passed to authoring service

bheyser committed:
removed identifier factory

bheyser committed:
cleanup after id rechange to former new idea

bheyser committed:
completed author and package in phpdoc class header

bheyser committed:
small fix

bheyser committed:
Answer stuff for cool playing

bheyser committed:
Answer stuff for cool playing

bheyser committed:
updated readme

bheyser committed:
example links for readme file

bheyser committed:
example links for readme file

alex40724 commented:
Thanks! I had a quick look at the readme. I have lots of questions, but start with only one, since this may answer others, too:

How would a consumer create a new question of a certain type? How would the consumer code look like?

bheyser committed:
nearly completed test player example

bheyser committed:
most close nearly completed test player example (docu remaining)

bheyser committed:
documentation

bheyser committed:
finalized documentation for example

bheyser committed:
pool shows create question button with link provided by the assessment question sevice, all creation is done within the service by the authoring command class

mstuder commented:
> How would a consumer create a new question of a certain type? How would the consumer code look like?

Hi Alex

Thx for your question!  As consumer you will not create a question, you will ask the Assessment Question Service for a creation link to the creation form:

```
$authoringService = $DIC->assessment()->service->authoring(
    $DIC->assessment->specification()->authoring(
        $myObjId, $myActorUserId, $myBacklink
    ),
    $DIC->assessment->consumer()->newQuestionUuid()
);	
$creationLinkComponent = $authoringService->getCreationLink();
```

The creation of the new question is done within the standard ILIAS Ctrl-Flow of the question service itself after post the creation form.

Best regards
Björn & Martin

bheyser committed:
updated readme

bheyser committed:
fixed method signature

mstuder committed:
- fixed missing creation link at the contract interface;
- readme enhancement

mstuder committed:
added more documentation to the readme file

mstuder committed:
fixed some typos

mstuder committed:
rearanged readme

mstuder committed:
fixed example

mstuder committed:
fixed example

mstuder committed:
fixed example

mstuder committed:
fixed readme examples

mstuder commented:
Referring to the open questions we enhanced the readme and hope, that this will help for the future discussion!

mstuder committed:
adapted UpdateUserAnswer for security reasons

mstuder committed:
revision readme

alex40724 commented:
@mstuder Thanks for the additional examples! This makes things clearer now.

Next question: Do you plan to use any value object types when data is being queried? E.g. when a consumer wants to display a list of questions:

https://github.com/ILIAS-eLearning/ILIAS/pull/2016/files#diff-b9e78755ef494819020e8979581ce378

It seems that you get associative arrays with no typisation which could be a weak point.

mstuder committed:
Implemented DTO example https://github.com/ILIAS-eLearning/ILIAS/pull/2016#issuecomment-515455979

mstuder commented:
@alex40724 
Do you mean something like this?
https://github.com/ILIAS-eLearning/ILIAS/pull/2016/commits/a9d7e80d3cfcd9d91bd026dad2930efb569895f9

Björn and I decided against it in a first discussion. Our thinking was that the API only contains things that are needed by the consumer. Test and Question Pool require AssocArrays for the current table UIs. Therefore we omitted this variant deliberately - I even think we have deleted the implementation of this method:-D

But, I also think that it might make sense, regardless of a current consumer demand.

mstuder committed:
Querying List fo DTO's in Readme

alex40724 commented:
Yes, current table UI very much relies on arrays. Thanks for the explanation.

mjansenDatabay commented on Services/AssessmentQuestion/examples/class.exTestPlayerGUI.php:
```diff
@@ -129,68 +117,112 @@ public function submitSolution()
+                    ),
+                    $user_id,
+                    json_encode(
+                        new PostDataFromServerRequest($request)->get('user_answer')
```
So this implies that all question types (and plugin question types) provide the data in `$_POST['user_answer']`, and the question type itself is responsible for the data and structure of data in `$_POST['user_answer']`?

What if there are multiple questions on one page (not relevant for the current state of the test object in ILIAS, but relevant in future).

mstuder replied:
> ... and the question type itself is responsible for the data and structure of data in $_POST['user_answer']

Yes!

> What if there are multiple questions on one page (not relevant for the current state of the test object in ILIAS, but relevant in future).

Thx for this question! Good Point!! 

I see two variants

Variant 1:
The consumer nevertheless sends the answers individually.  He could place a try/catch block outside a foreach. For this, however, the POST variable by rendering the question would have to be an array already, i.e. user_answer[].

In this case everything would be stored except those that fail.

Variant 2:
You only want to save the page if the whole page is saved. In this case we would have to offer another communication interface Post**Array**DataFromServerRequest.

Are there other variants? Should we already now provide the user_answer in the representation as an array?

mjansenDatabay replied:
If ILIAS supports multiple questions on a page in future, a `Transaction` like handling (begin, commit, rollback) of the/all user answers could be maybe a requirement. So a 'everything succeeds or fails' behaviour could be the variant we should support.

But IMO both variants could be a valid scenario for the consumer of the question service. 

For the HTTP communication between (user) client and the consumer endpoint (e.g. UI controller of the Test repository object) the UUID of the question should be part of the structure of the `$_POST['user_answer']` array, e.g.:

```php
$_POST['user_answer']['d01a14f5-544d-418f-87cf-7c293c673531'] = [
    'gap1' => 'a² + 2ab + b²', // Some question specific user answer data
];

$_POST['user_answer']['90f5f1f7-f3ab-456e-940c-66006d99c1de'] = [
    'essay' => 'ILIAS is a powerful LMS ...', // Some question specific user answer data of another question
];
```

bheyser replied:
In addition to Martins variants for solving the problems with the user_answer post variable,
Variant 3:

The consumer will define the post variable to be used by the question types. This way the consumer gets the full control over his own form fields' post variables.

bheyser replied:
In general it will be technically possible to implement either a user answer save service that takes all user answers at once or a service that takes care about one user answer only.

The question is from high value, we will consider this in mind. I just need to say, the question also aims general conceptional questions, that needs to be clarified within the related feature request.

Even the offline implementation of questions supporting the test and also the learning module could have impact on this question.

Very Thanks @mjansenDatabay  (!) <3

mjansenDatabay commented on Services/AssessmentQuestion/examples/class.exTestPlayerGUI.php:
```diff
@@ -129,68 +117,112 @@ public function submitSolution()
+			$this->buildAsqPlayServiceSpec(), $DIC->assessment()->consumer()->questionUuid($questionUuid),$DIC->assessment()->consumer()->revisionUuid($questionUuid,$revisionUuid)
+		);
+		
+		$DIC->assessment()->service()->play()->CreateUserAnswer(
```
What I don't get (yet) by reading the examples and the README file is the relation between `CreateUserAnswer` and `GetUserScore`.
`CreateUserAnswer` does not return or process anything, yet. What exactly is the purpose of this method? Does it store the raw particpant data? Or is the consumer itself responsible for storing the data based on a (maybe) future return value of `CreateUserAnswer`?
The next interesting command is `GetUserScore`, which returns an `UserAnswerScoringContract` instance which can be used for result calculcation and storage. This method seems to be independent of `CreateUserAnswer`, so I guess it will always get the score from the persistence layer, right?

Please provide a few more sentences according to result storage and calculation in the README file.

mstuder replied:
> CreateUserAnswer does not return or process anything, yet. What exactly is the purpose of 
> this method?


In the past, _user answer_ was called _user solution_. But my staff and I simply can't use this term. While a solution is the correct answer for us. :-D

CreateUserAnswer saves the answer given by a user. This must be a new answer to the question. If a previously given answer is to be updated UpdateUserAnswer must be chosen.

GetUserScore returns the score for the given answer from the point of view of the question sercice. The consumer specific settings like _For Each Questions Negative Points are set to '0 Points'_ are not considered here. This is a matter for the consumer.

I will update the Readme. There are errors in the proposal also.

CreateUserAnswer(...):int -> CreateUserAnswer(..):void

mstuder replied:
I have done the updates https://github.com/ILIAS-eLearning/ILIAS/blob/e73fe5ea8692febf11c27b5f9262dd1d4047d4f9/Services/AssessmentQuestion/docs/readme.md

Thx for your questions and your time!

mjansenDatabay replied:
> In the past, _user answer_ was called _user solution_. But my staff and I simply can't use this term. While a solution is the correct answer for us. :-D

I totally agree.

> CreateUserAnswer saves the answer given by a user. This must be a new answer to the question. If a previously given answer is to be updated UpdateUserAnswer must be chosen.
GetUserScore returns the score for the given answer from the point of view of the question sercice. The consumer specific settings like For Each Questions Negative Points are set to '0 Points' are not considered here. This is a matter for the consumer.

Thx for clarification and for updating the README file.

mstuder committed:
optimized Readme for understanding user_answer and user_score
fixed bugs in example code and interfaces

mstuder committed:
Merge branch 'asq_api_change_proposal_discussion_r308090855' into 'feature/6-0/asq_api_change_proposal'

optimized Readme for understanding user_answer and user_score

See merge request ILIAS/Core!247

mstuder committed:
fixed readme

bheyser committed:
changed types

bheyser committed:
completed readme with offline standalone question export

mjansenDatabay commented:
Hi @mstuder,

thanks a lot for this PR, it really helps to clarify what the proposed changes in the interface to the new services are.

In general it was hard to see the essential difference between this proposal and Björn's initial proposal (see our first three questions). We currently do not see much difference in terms of expressive strength or even general structure, besides the usage of UUIDs. We see a lot of naming changes, new classes and interfaces, but it was hard to understand how these contribute to make this API implementable when Björn's initial proposal wasn't. It would really help to understand which changes are essential for this quality and which are not.

On the other hand, we do see a lot of room to improve the usability of this new API. The initial proposal from Björn seems to be a lot easier to understand and hence use than this proposal. This, however, just seems to be a matter of naming and some cleanup and tidying in the surface of the API. We left some suggestions that we feel will improve the usability of the API a lot. Feel free to ask @klees per skype or phone for further insights. We feel that none of the proposed changes will influence essential properties of the new API-design, but we will of course be happy about new insights here.

Since we currently do not see many essential changes w.r.t. to Björn's proposal, we think we can move on with this proposal once the usability is improved and the changes are implemented. It is very understandable that Björn's initial proposal needs improvement, details tend to crop up once the implementation is started. We hope this can be done while maintaining the usability of the API for consumers. We still want to understand what triggered the changes in the first place because we feel that this also will inform the discussion about the proposed implementation.

Please answer the following questions:
- [x] Why did you change the public API in comparison to the approach Björn proposed in PR https://github.com/ILIAS-eLearning/ILIAS/pull/1183 ?
- [x] What are the benefits of these changes?
- [x] Which impediments did you and your team experience with the old API proposal?
- [x] What are these `EventSubscriber` in `AuthoringServiceSpec`? They are nowhere to be found in the PR.
- [x] How do you imagine to implement the "Corrections Mode" (Nachkorrektur) with this or are you planing to implement all of it in the player?


Please consider the following suggestions. You do not need to follow those, but please indicate shortly why you prefer to do otherwise:
- [x] Please remove the `Spec`-classes and -interfaces from the public interface completely. Instead pass the required arguments directly. `Specs` are always consumed immediately.
- [x] Please remove the "command"-classes (`UserAnswerSubmitContract`, maybe there are more?) from the public interface. Instead pass the required arguments directly. The class(es) are always consumed immediately.
- [x] Please remove the `interfaces` for data-classes like `IDs`. There does not seem to be the requirement to implement them differently or hide implementation details.
- [x] Please reconsider the naming. It is highly unconventional for ILIAS (consider `Spec` and `Contract`) and partly hard to understand. E.g.: Why do we have a `PlayService` that only deals with one Question? This could just be a `QuestionPlayer`, right? Consider the snippet you show in the README at "Get the Service". Why are some of the factories in `consumer` and others in `specification`?
- [x] Please reconsider the use of static method calls. They are maybe an indicator of structural deficits and missing dependency management.
- [x] Please reconsider the creation of fresh UUIDs. Currently, the consumer needs to create them and thus needs to know how to do so. It seems to be easily possible to move the creation of new UUIDs inside the service and free consumers from the requirement to know how create them.

Please implement the following changes:
- [x] Follow our PSR-2 (+ additions) coding style guideline
- [x] Use strong typing (`declare(strict_types=1);`)
- [x] Add return type declarations and type hints (where missing)
- [x] Use `lowerCamelCase` method names instead of `UpperCamelCase` method names
- [x] Add `PHPDoc` documentation for the classes in `./Services/AssessmentQuestion/PublicApi/Factory`

We are looking forward to discuss this PR with you in our meeting!

Best regards,
The Technical Board

mstuder committed:
WIP Refactoring

mstuder committed:
Refactoring

mstuder committed:
Refactoring Readme

mstuder committed:
refactoring an reformatted

mstuder commented:
Dear @mjansenDatabay 
Dear TB

Thank you very much for this detailed feedback! It helped us to take a step back and further optimize our service!

> It would really help to understand which changes are essential for 
> this quality and which are not.

None of the changes are really necessary. In our case, the service is so detached from the actual implementation of the assessment questions that we can work with any service.

> This, however, just seems to be a matter of naming and some cleanup and tidying 
> in the surface of the API.

We have now further refined the proposal. Björn and I think that he will be able to use the PR #1183 first suggestion much better now. 

> This, however, just seems to be a matter of naming and some cleanup and tidying 
> in the surface of the API.

We have now further refined the proposal. Björn and I think that he will be able to use the PR #1183 first suggestion much better now. 

> We still want to understand what triggered the changes in the first place because 
> we feel that this also will inform the discussion about the proposed 
> implementation.

We find the proposal has a clearer structure and a better thematic outline than the PR #1183 approach. The proposal 1183 knew exactly one Factory $DIC->question()->getEverything(). We now provide a clear outline of authoring and processing. 

With the approach PR #1183 DomainEntities had to be handed over by the Assessment Question itself. There was no clear boundary between service and the underlying implementation. With this proposal we want to introduce a so-called Anti-Corruption Layer.

With this approach it is much clearer where new requirements will be placed. Our new approach is externally oriented to the provision of all use cases. The old approach was based on making the internal domain entities with the necessary interfaces understandable to the outside world so that the consumer can create use cases.

As I said, we could also work with the old approach. But please take a look at our revision now. We're sure you'll like it a lot better than PR #1183.

> Please answer the following questions:
> Why did you change the public API in comparison to the approach Björn proposed 
> in PR #1183 ?

Because we want to draw a clear line between service and our domain.

> What are the benefits of these changes?

The better understanding and the clearer process flow. For example, we don't have to return any return values for "Commands". The consumer doesn't have to put himself on a listening list either. By generating the UUID from outside, the complexity is considerably reduced.

> Which impediments did you and your team experience with the old API proposal?

None. We didn't have any trouble with the service. We had trouble with the decision that all questions would be extracted 1:1 to the test and question pool and that the over 10 year old code should have been used 1:1. A thing of impossibility to implement this with as few bugs as possible.

> What are these EventSubscribers in AuthoringServiceSpec? They are nowhere to be 
> found in the PR.

We have removed them. We don't think that we will need them.  

> How do you imagine to implement the "Corrections Mode" with 
> this or are you planing to implement all of it in the player?

Certainly on the processing side. But here with a separate GUI and separate service.

> Please remove the Spec-classes and -interfaces from the public interface 
> completely. Instead pass the required arguments directly. Specs are always 
> consumed immediately.

Done. 

> Please remove the "command"-classes (UserAnswerSubmitContract, maybe there 
> are more?) from the public interface. Instead pass the required arguments directly. > The class(es) are always consumed immediately.

Done.  

> Please remove the interfaces for data-classes like IDs. There does not seem to be 
> the requirement to implement them differently or hide implementation details.
> Please reconsider the naming. It is highly unconventional for ILIAS (consider Spec 
> and Contract) and partly hard to understand. 

With Contracts we have thought that we could provide appropriate guarantees to the outside world and that later we could do something like service versioning easier. Release version 2 of the contract and the service will implement both contract interfaces for a while.

We have removed these and are waiting for the proposal of the Api Alignment Group, which will obviously provide answers here.

> E.g.: Why do we have a PlayService that only deals with one Question? This could 
> just be a QuestionPlayer, right? Consider the snippet you show in the README at 
> Get the Service. Why are some of the factories in consumer and others in 
> specification?

We have optimized it.

> Please reconsider the use of static method calls. They are maybe an indicator of 
> structural deficits and missing dependency management.

We didn't find any.

> Please reconsider the creation of fresh UUIDs. Currently, the consumer needs to 
> create them and thus needs to know how to do so. It seems to be easily possible to > move the 
> creation of new UUIDs inside the service and free consumers from the 
> requirement to know how create them.

From our point of view it is important and right to let the consumer do this. In this way, he can store data for the questions that are consumer-specific even before the questions are generated. No return values and no event listeners are required. It is a clear concept if the UUIDs are demanded by the consumer everywhere and not only at selected places.

> Follow our PSR-2 (+ additions) coding style guideline

done.
 
> Use strong typing (declare(strict_types=1);)

done

> Add return type declarations and type hints (where missing)

done

> Use lowerCamelCase method names instead of UpperCamelCase method names

done

Thx for your time and your quality awareness. We are looking forward to hearing from you.

Best regards
Björn & Martin

mstuder committed:
fixed some Readme 'Bugs'

mstuder committed:
fixed some Readme 'Bugs'

swiniker commented:
Dear @mstuder,

thanks a lot for this smaller PR. This really helps to clarify, how you want to introduce the proposed patterns for the Assessment Question Service. Please note, that this PR still is very hard to review and it took every member of the TB considerable time to do so. Besides the size of the changes and the variety of patterns (which simply might not be reducible or divisible) we found a lot of open ends (like missing classes or implementations) in the PR. This led to a lot of open questions and lack of clarity. Also, this PR seems to show only some of the implications and complexities of proposed patterns. Your README helps a lot in this regard, but in the end, the truth is in the code here.


Since this is about the patterns you want to introduce in the service, we like to start by giving feedback regarding these.

**We support the introduction of UUIDs as IDs** instead of using database generated incremented IDs. This move will allow to decouple the persistence from the creation of objects.

**We do not object to introduce an *internal* command bus in the service.** Since we only see rudimentary pieces of the command bus and the implications of it for the implementation and structure of the service, we are not sure how this will impact the service. We understand how a command bus could help to structure and ease the implementation of services in general. If you feel that a command bus will help to solve problems in the Assessment Question Service, feel free to use it internally. We want to stress again, though, that a command bus is a major consideration for the Internal API Alignment team when thinking about a centralized API handling. Introducing a command bus here implies the possibility of at least two similar approaches (maybe on another architecturel layer). We expect you to refactor your `CommandBus`-related code accordingly if such concepts become part of the ILIAS core after the approval of the Internal API Alignment concept and if there is an intersection between the target layers of those. We gladly acknowledge that you already agreed to do so. Thanks! Also thanks for tackling this topic and sharing the acquired knowledge with the rest of the community in discussions such as this one.

**We strongly advise against the introduction of event sourcing while factoring out the Assessment Question Service from the T&A.** We do see that this approach could have a general value for the Assessment Question Service as well as for other components in ILIAS. We understand which new features of the Assessment Question Service could benefit from the approach. However, currently the absolut priority for the T&A (which we clearly communicated several times) is to factor out an Assessment Question Service from the T&A. We cannot see how the event sourcing works towards that goal. We also failed to find a traceable explanation why the introduction of event sourcing is imperative to make the implementation of the API as outlined in your/Björns proposal possible. Instead, **we are heavily concerned about the tremendous amount of additional complexity and hence sources for bugs and various other issues** (e.g. with performance) that this approach will invariably introduce. We are pretty sure, that this PR only shows a small part of the real complexity that will be involved with event sourcing. Currently the PR does not answer the following questions (among others):

* What will the normalized database scheme look like that will exist besides the event-sourced database?
* How will the data be moved from the event store to the normalized tables? When is the right moment in time for doing so?
* How will you deal with currently existing questions (incl. plugins) and how will they be migrated to the event-sourced database?
* How will the event sourcing mechanism deal with schema changes when the data is stored in json-strings?
* How will the event sourcing mechanism deal with binary data?
* Which additional features (regarding this PR) for the event sourcing are planed for ILIAS 6.0? The normalization, e.g., currently is not included here?

All answers to these questions will introduce new complexity. We think the approach could very well be considered when the Assessment Question Service moves on after it is separated from the rest of the T&A, but we feel that this separation alone introduces enough difficulties for now. 

**We strongly advise against the introduction of CQRS.** We see the general value of that approach, but we also see that there are considerable downsides too (see e.g. [Martin Fowler](https://martinfowler.com/bliki/CQRS.html)). We currently do not feel that CQRS is a good fit for the Assessment Question Service or the system in general, considering the pros and cons of the approach. As always, this is up for debate in general, but we do not think that the seperation of an Assessment Question Service from the T&A is a good place for this debate. Further, it is hard to see how this pattern is really show cased in the PR, as we cannot see different models for read and write. In the current state of the work, we can't tell which parts of the PR are really about CQRS as we see a mix of many patterns and details of the implementation. Maybe it's just best to drop the label 'CQRS' and focus on the concrete requirements of the Assessment Question Service and the implementation. Same as for event sourcing, we do not see a tangible explanation, why the implementation of the target APIs of the service is only possible or even just considerably simplified when using CQRS and **we are deeply concerned about the complexities that CQRS will introduce now and in the future for every involved party.** We see how the new feature [Item Statistics](https://docu.ilias.de/goto_docu_wiki_wpage_5321_1357.html) could lead to the idea to introduce CQRS but, again, we clearly prioritized the separation of the Assessment Question Service from the rest of the T&A over new features. We propose that you consider a [Reporting Database](https://martinfowler.com/bliki/ReportingDatabase.html) as a simpler alternative to CQRS. This could also be introduced after the separation.

 **We definitely see that we as a community should investigate DDD further, consider ideas as well as processes and try to learn lessons from DDD. Thanks for bringing it up.** You are implementing DDD by introducing entities, services, application as e.g. presented in "Domain Driven Design in PHP" by Buenosvinos. However, note that Buenosvinos and others also describe DDD as an approach to design a system in a close cooperation between domain experts and developers by creating and documenting a shared model, the ubiquitous language. The implementation then uses certain techniques to implement that language, but the approach itself is not about the implementation only but also about the design process. It is no problem, though, that this design process was not followed for the separation of the Assessment Question Service, as there already exist requirements and models and there certainly was no time to go through an intense iterative DDD process.

Regarding the other content of the PR, we want to add the following suggestions and requirements:

* Please use dependency injection instead of static method calls. The massive use of static method calls indicates a lack of code organisation and structure.
* We strongly recommend to introduce automated tests. This will be simplified by introducing dependency injection.
* Use the mandatory [PSR-2 (+X) ILIAS Coding Style](https://github.com/ILIAS-eLearning/ILIAS/pull/1838/files).
* Add return types and type hinting where possible.

**We want to restate publicly, that we just try to play our role as advisory body for the maintainers and the product manager here. Ultimately it is up to you as maintainers now to decide how to go on with this (also concerning the question of whether to use or not to use CQRS and Event Sourcing).** We are trying hard to find a course of action that allows for a consensual way forward here. To balance your agency as maintainers with our deep concerns regarding your proposal, we propose the following measures:

* Create a branch in the public ILIAS-repo where everyone (we, your customers, other developers, ...) can see the current state of the implementation. This will increase transparency and also clarify where we stand in terms of functionality and general progress. We realized that we currently do not understand how far you really have progressed in the separation of the Assessment Question Service which might influence our assessment of the situation.
* Ask Matthias to create a test-installation for that branch that people can access. It will help to alleviate our concerns if we could see your concepts in action. This will also allow everyone to test the new implementation to make sure that bugs are found early and to show that your proposal in fact works despite our reservations. This will also allow to go on with ILIAS 6.0 without the changes for the assessment question service if our concerns turn out to be warranted.
* Both measures will require a transparent and ongoing communication on your side to manage expectations towards the current state of the implementation.
* **Important:** Note, that the branch can only be merged to the trunk when approved by your customers. The TB is among these customers as the sponsor of the initial concept for the separation of the service. Note also, that there are other improvements of and investments in ILIAS 6.0 at stake here as well. We will not agree to merge any state of the branch that puts these other improvements and investments at an unknown or uncontrollable risk. 

We hope that we clearly and unambiguously pointed out where and why we are concerned about the patterns you want to introduce here. Please, don't hesitate to ask for clarifications, if we didn't succeed as well as we thought. We might be wrong, but we cannot see how this can further be decide based on some abstract discussion at this point. Please appreciate our concerns, however you may decide to go on. We do not try to stop you for no good reason, but we have a responsibility towards the whole system, given by the by-laws of the ILIAS society.

We are looking forward to our meeting and hope we find a way forward together.

Kind regards,
the Technical Board

Amstutz commented:
We are not sure, what the relation to: https://github.com/ILIAS-eLearning/ILIAS/pull/2121 is? We guess that this one here can be closed, correct?

mstuder commented:
This pull request was used to discuss new technological approaches within the assessment question service. The concrete development you find under https://github.com/ILIAS-eLearning/ILIAS/pull/2121
